File: 4.8-bc/ts/stage-2-feature-first/src/features/auth/handler.ts
import type { FastifyInstance } from "fastify";

import { writeError } from "../../shared/httputil.js";
import { InvalidCredentialsError, Service } from "./service.js";

interface CredsBody {
  email?: string;
  password?: string;
}

export class Handler {
  constructor(private readonly svc: Service) {}

  register(app: FastifyInstance): void {
    app.post<{ Body: CredsBody }>("/auth/register", async (req, reply) => {
      const { email, password } = req.body ?? {};
      if (!email || !password) {
        return reply.code(400).send({ error: "invalid_request" });
      }
      try {
        const u = await this.svc.register(email, password);
        return reply.code(201).send({ user_id: u.id, email: u.email });
      } catch (err) {
        req.log.error({ err }, "register failed");
        return writeError(reply, err);
      }
    });

    app.post<{ Body: CredsBody }>("/auth/login", async (req, reply) => {
      const { email, password } = req.body ?? {};
      if (!email || !password) {
        return reply.code(400).send({ error: "invalid_request" });
      }
      try {
        const u = await this.svc.login(email, password);
        return reply.code(200).send({ user_id: u.id, email: u.email });
      } catch (err) {
        if (err instanceof InvalidCredentialsError) {
          return reply.code(401).send({ error: "invalid_credentials" });
        }
        req.log.error({ err }, "login failed");
        return writeError(reply, err);
      }
    });
  }
}

File: 4.8-bc/py/stage-2-feature-first/app/features/auth/handler.py
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel, EmailStr
from sqlalchemy.ext.asyncio import AsyncSession

from app.features.auth.repository import Repository
from app.features.auth.service import InvalidCredentialsError, Service
from app.shared.db import get_session

router = APIRouter(tags=["auth"])


class CredsReq(BaseModel):
    email: EmailStr
    password: str


class UserResp(BaseModel):
    user_id: str
    email: str


@router.post(
    "/auth/register", status_code=status.HTTP_201_CREATED, response_model=UserResp
)
async def register(
    req: CredsReq, session: AsyncSession = Depends(get_session)
) -> UserResp:
    if not req.password:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST, detail="invalid_request"
        )
    svc = Service(Repository(session))
    u = await svc.register(req.email, req.password)
    return UserResp(user_id=str(u.id), email=u.email)


@router.post("/auth/login", response_model=UserResp)
async def login(
    req: CredsReq, session: AsyncSession = Depends(get_session)
) -> UserResp:
    svc = Service(Repository(session))
    try:
        u = await svc.login(req.email, req.password)
    except InvalidCredentialsError as exc:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED, detail="invalid_credentials"
        ) from exc
    return UserResp(user_id=str(u.id), email=u.email)
